docs/NNNN-n-tuple-omit-prefix-storage-layout.md

```diff
@@ -0,0 +1,183 @@
+The object id is then divided into N n-tuple segments, and used to create nested paths under the OCFL storage root, followed by the prefix-omitted object id directory.  
+
+This layout combines the advantages of 0006-flat-omit-prefix-storage-layout (directory name transparency) and the 0004-hashed-n-tuple-storage-layout (enhanced file system/bucket performance). 
+
+The limitations of this layout are filesystem dependent, but are generally as follows:
+
+* The size of object identifiers, minus the length of the prefix, cannot exceed the maximum allowed directory name size
+  (eg. 255 characters)
+* Object identifiers cannot include characters that are illegal in directory names
+
+## Parameters
+
+### Summary
+
+* **Name:** `delimiter`
+  * **Description:** The case-insensitive, delimiter marking the end of the OCFL object identifier prefix; MUST consist
+    of a character string of length one or greater. If the delimiter is found multiple times in the OCFL object
+    identifier, its last occurence (right-most) will be used to select the termination of the prefix.
+  * **Type:** string
+  * **Constraints:** Must not be empty
+  * **Default:** :
+* **Name**: `tupleSize`
+  * **Description:** Indicates the segment size (in characters) to split the
+    identifier into
+  * **Type:** number
+  * **Constraints:** An integer between 1 and 32 inclusive
+  * **Default:** 3
+* **Name:** `numberOfTuples`
+  * **Description:** Indicates the number of segments to use for path generation
+  * **Type:** number
+  * **Constraints:** An integer between 1 and 32 inclusive
+  * **Default:** 3
+* **Name:** `zeroPadding`
+  * **Description:** Indicates whether to use left or right zero padding for ids less than tupleSize * numberOfTuples
+  * **Type:** string
+  * **Constraints:** Must be either "left", "right" or "none"
+  * **Default:** left
+* **Name:** `reverseObjectRoot`
+  * **Description:** When true, indicates that the prefix-omitted, padded object identifier should be reversed
+  * **Type:** boolean
+  * **Default:** false
+
+## Procedure
+
+The following is an outline of the steps to map an OCFL object identifier to an
+OCFL object root path:
+
+1. Remove the prefix, which is everything to the left of the right-most instance of the delimiter, as well as the delimiter. If there is no delimiter, the whole id is used; if the delimiter is found at the end, an error is thrown.
+2. Optionally, add zero-padding to the left or right of the remaining id, depending on `zeroPadding` configuration.
+3. Optionally reverse the remaining id, depending on `reverseObjectRoot`
+4. Starting at the leftmost character of the resulting id and working right, divide the id into `numberOfTuples` each containing `tupleSize` characters.
+5. It is expected that the UTF-8 identifier from step 5 is split between grapheme clusters
```

Comment:
Step 5 isn't a step, it is a comment. Add to end of 4 if to be kept. Also, introducing the word "grapheme" seems confusing rather than clarifying

Reply:
I made step 5 a comment on step 4 instead and fixed numbering. Regarding mention of "grapheme", this was from an earlier suggestion, but I'm open to modifications that would make this clearer ...

Reply:
I added this paranthetical, lmk what you think:
(e.g. character sequences such as (g + °),  digraphs such as Slovak “ch”, etc. - see more at https://www.unicode.org/reports/tr29/tr29-1.html#Grapheme_Cluster_Boundaries).

Reply:
Note the reference to "step 5" in the now combined step 4 needs to be fixed.

I'm still not really sure what the implication of the mention of grapheme is. Let's say that I have an identifier:

`abg̊defg`

where the g̊ is (g + °). What is the expected behavior with n=3 and N=2? Is it:

* `abg̊/def/abg̊defg` (technically 4 chars then 3 chars, splitting on a grapheme boundary)
* `abg/ ̊de/abg̊defg` (3 chars and 3 chars, breaking a grapheme)

If the intention is the former then I think the whole notion of grapheme needs to be introduced earlier and essentially the n-tuple must be defined differently.

Reply:
I agree that the grapheme discussion is confusing, I've removed references to grapheme clusters altogether for now, but please let me know if I should revisit in some way.

Reply:
Removing the language does not remove the problem. How does this extension handle @zimeon's example?

Reply:
I think that if (as now) written solely in terms of characters the answer is `abg/ ̊de/abg̊defg` (3 chars and 3 chars, breaking a grapheme)

Reply:
To me, that seems undesirable. Are others okay with that?

Reply:
Pro: well defined and can be reliably implemented
Con: non-intuitive split of grapheme in path elements which also makes the path appear not to match the identifier

Godo question. I think, on balance, I would be happier to stick with character semantics. The thought of implementations paying attention to n-grapheme-tuples (algorithm in https://www.unicode.org/reports/tr10/#Collation_Graphemes ) seems risky

Reply:
@zimeon Did you create your example output by hand or by code? I'm having a hard time reproducing it. In Java and Rust I get the following when I split by character: `abg/̊de/abg̊defg`

Reply:
Okay, I created both directory structures, and on my system it is impossible (or I couldn't figure out how to make it work) to navigate into the directory with the split grapheme via the terminal (works in a gui) and it displays the directory simply as `de`.

Personally, I prefer splitting on grapheme clusters because it is the easiest for humans to use. Splitting on characters is the easier for machines.

Reply:
@pwinckles - my earlier suggestions were entirely theoretical.

I have now checked in code. My terminal seems to happily combine the ring character with a slash so things get a bit ugly in display (better in this web window). I verified the following:

```
(py38) simeon@RottenApple ~> cat p.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
id = u'abg̊defg'
N = 2
n = 3

segs = []
for t in range(0, N):
    segs.append(id[t * n:(t + 1) * n])
segs.append(id)
print('/'.join(segs))
(py38) simeon@RottenApple ~> python p.py
abg/̊de/abg̊defg
(py38) simeon@RottenApple ~> mkdir abg
(py38) simeon@RottenApple ~> mkdir abg/̊de
(py38) simeon@RottenApple ~> touch abg/̊de/abg̊defg
(py38) simeon@RottenApple ~> ls -l abg
total 0
drwxr-xr-x  3 simeon  staff  96 Nov 10 14:28 ̊de
(py38) simeon@RottenApple ~> ls -l abg/̊de
total 0
-rw-r--r--  1 simeon  staff  0 Nov 10 14:28 abg̊defg
```

It does seem that there is package for working on graphemes in Python (https://pypi.org/project/grapheme/) that can be substituted to get grapheme splitting:

```(py38) simeon@RottenApple ~> cat g.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import grapheme
id = u'abg̊defg'
N = 2
n = 3

segs = []
for t in range(0, N):
    segs.append(grapheme.slice(id, t * n, (t + 1) * n))
segs.append(id)
print('/'.join(segs))
(py38) simeon@RottenApple ~> python g.py
abg̊/def/abg̊defg
```
but I don't know how good that is or how well implemented on other languages grapheme semantics are

Reply:
I would assume that it's fairly common.

Java: https://icu.unicode.org/
Rust: https://crates.io/crates/unicode-segmentation
Javascript: https://github.com/orling/grapheme-splitter
Go: https://github.com/rivo/uniseg

...

Reply:
I played around with your example on my machine. It looks like the trick to make it work, at least for me, is the inclusion of the leading `/`. So, I can `cd abg/̊de` or `cd ./̊de`, but can't if it's just the directory name

Reply:
I can `cd   ̊de` just fine (on mac & linux vm)

Reply:
Yes... that does work for me. My method was to copy the string and then delete the part at the beginning. It seems like the `°` is associated to the `/` and is deleted when the `/` is removed.

Reply:
Thanks for the discussion, and sorry I simply removed the grapheme reference. I'm open to any suggestion that you think makes sense, and I can talk to Andrew as well. FYI, I'll be out until Monday, but I will also occasionally be checking in, so I can move this forward if there's some resolution. Or we can pick up Monday if you'd like any additional Harvard input.

Reply:
@michael-lts - thanks for you patience with us! I think it is much more important to get a good resolution than to do it quickly, enjoy your time out!

Reply:
In terms of desired behavior, I lean towards the previously expressed [sentiment](https://github.com/OCFL/extensions/pull/52#discussion_r746850165); meaning, splitting graphemes is undesirable:
>  @zimeon 5 days ago
> I think that if (as now) written solely in terms of characters the answer is abg/ ̊de/abg̊defg (3 chars and 3 chars, breaking a grapheme)
> 
> @pwinckles pwinckles 5 days ago
> To me, that seems undesirable. Are others okay with that?

Is that a shared perspective? It would be good to agree on the goal.

Additionally, it appears that there is language support for handling graphemes, which presumably allows for managing them as single units.

Reply:
Hello all, we are leaning toward a statement that this extension does not define the behavior of graphemes. If that is acceptable, is there any specific wording you'd like to see (as a comment on #4 fo the Procedure section?). Is it worth mentioning any language support, or pointing to any documentation? Thanks!

Reply:
Personally, I am not crazy about ambiguous behavior. If you don't want to address the problem, I would just as soon state in the extension that it's only intended to support ascii.

Reply:
@pwinckles : Downgrading to ASCII seems a little extreme, no? Do you feel like that is necessary in order to defer on the topic of graphemes?

Reply:
I don't think so. The problem is not a rare edge case for non-English languages. The behavior would need to be defined in order to have reproducible results across implementations. Otherwise, what's the point of adopting a specification?

Reply:
Hi, I just pushed a new version with a comment that only ascii is supported, please let me know if you'd like me to tweak it in any way, thanks

Reply:
@zimeon what are you current thoughts on the character problem?

Reply:
Re suggestion to call out ascii issue in overview, I went ahead and did that, good suggestion.